### ISSUE TITLE
Early stopping in a Scikit Pipeline

### ISSUE THREAD

**Issue by mratsim** [dmlc/xgboost#2039]
I use XGBClassifier as the last step of a Pipeline which works great.

However I am unable to use early stopping.
Early stopping needs "eval_set" to be set, but there is no way to easily reapply the pipeline transformation to eval_set.

For example, this code would not work because XGBClassifier does not implement the transform method:
```python
eval_set = [(pipe.transform(X_test),pipe.transform(y_test))]
```

I see two potential solutions to this:

1. Implement a dummy transform method which according to this [Scikitlearn documentation](http://scikit-learn.org/stable/developers/contributing.html#pipeline-compatibility) will allow to apply the Pipeline to arbitrary data. 

2. Similar to [Keras](https://keras.io/models/sequential/), implement a validation_split parameter that  holds out part of the input data to be used as eval_set.

I raised the same issue for LightGBM - https://github.com/Microsoft/LightGBM/issues/299

**Comment by mratsim** [comment]
From discussion in linked LightGBM issue

----

## For solution 1  - Implement a dummy `transform` method:
- This would be a workaround upon a lack of functionality in `sklearn` Pipeline API. Most classifiers even in sklearn do not have a `transform`. And for those that do have (Decision Trees and variants), it was used to extract features and is deprecated as you linked.
Adding a dummy `transform` method would be make it very confusing for people on what to expect (dummy, feature extraction, ...)

## For solution 2 - Implement a validation split like in `Keras`:
- The issue here is the lack of control on the validation set and the split. I If there is a need to inspect deeper into validation issues, I suppose it would be non-trivial to extract it from the classifier or provide an API for it.


My recommendation would be for a Sklearn.pipeline method to apply all pipeline transformations to an arbitrary dataset (validation set usually).
Hence, I will open a ticket to `Scikit-learn` and close this ticket (unless further discussion) in the next 3 days.

**Comment by mratsim** [comment]
Closing. Issue raised upstream to Scikit-learn
